src/utilities.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Utilities used by the charm."""
+
+import logging
+import subprocess  # nosec B404
+import time
+from typing import Callable, Optional, Sequence, Type, TypeVar
+
+logger = logging.getLogger(__name__)
+
+# Return type of decorated function
+R = TypeVar("R")
```

Comment:
Would it be worth expanding on the variable name? Maybe `TReturn`? I think there is some convention, perhaps `mypy` enforces it